homeassistant/components/media_player/cambridgeaudio.py

```diff
@@ -0,0 +1,389 @@
+        self._smdevice.trnsprt_seek(abs_pos)
+
+    def media_stop(self):
+        """Stop playing the current media."""
+        self._smdevice.trnsprt_stop()
+        self._state = STATE_IDLE
+
+    def media_play(self):
+        """Start playing the current media."""
+        self._smdevice.trnsprt_play()
+        self._state = STATE_PLAYING
+
+    def mute_volume(self, mute):
+        """Mute the volume. This only works in pre-amp mode."""
+        self._smdevice.volume_mute(mute)
+
+    def play_media(self, media_type, media_id, **kwargs):
+        """Play the specified media."""
+        pass
+
+    def select_source(self, source):
+        """Switch to the selected internet radio preset."""
+        self._smdevice.play_preset(self._sources_reverse[source])
+        self.schedule_update_ha_state()
+
+    def set_shuffle(self, shuffle):
+        """Enable/disable shuffle play."""
+        self._smdevice.set_shuffle(shuffle)
+
+    def set_volume_level(self, volume):
+        """Set the current volume level. This only works in pre-amp mode."""
+        volume = int(float(self._volume_max) * float(volume))
+        self._smdevice.set_volume(volume)
+
+    def turn_off(self):
+        """Switch the device OFF (default) or into IDLE mode."""
+        _LOGGER.debug("Powering off (%s)", self._power_off_cmd)
+        self._smdevice.power_off(power_state=self._power_off_cmd)
+        self._pwstate = STATE_OFF
+
+    def turn_on(self):
+        """Turn the device on if it's in IDLE (network standby) mode."""
+        if self._smdevice.get_power_state() == 'idle':
+            self._smdevice.power_on()
+            self._pwstate = STATE_ON
+
+    @staticmethod
+    def _to_seconds(timespec):
+        """Helper function to translate a h:mm:ss time spec into seconds."""
+        hours, minutes, seconds = timespec.split(':')
+        return int(hours) * 60 + int(minutes) * 60 + int(seconds)
+
+    def update(self):
+        """Fetch new state data from the media_player."""
+        dev = self._smdevice
+
+        pwstate = dev.get_power_state()
+        if pwstate is None or pwstate in ['idle', 'off']:
+            self._pwstate = STATE_OFF
+            self._state = STATE_OFF
+            return
+        self._pwstate = STATE_ON
+        self._state = {
+            'PLAYING': STATE_PLAYING,
+            'PAUSED_PLAYBACK': STATE_PAUSED,
+            'STOPPED': STATE_IDLE,
+            'TRANSITIONING': STATE_PLAYING}.get(dev.get_transport_state())
+        self._audio_source = dev.get_audio_source()
+
+        try:
+            self._source = dev.get_current_preset()['name']
+        except TypeError:   # in case get_current_preset() returns None
+            self._source = None
+
+        # only need to update volume if the device supports changing it
+        if self._support_volumecontrol:
+            self._muted = dev.get_mute_state()
+            self._volume = dev.get_volume()
+
+        if self._state == STATE_PLAYING:
```

Comment:
Consider using a guard clause to keep the indentation sane:

```python
if self._state != STATE_PLAYING:
    return

if self._audio_source == "media player":
    …
```

Reply:
Done, thanks!